Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,326 @@
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ) : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relations corresponding to a Coxeter matrix. -/
+@[nolint unusedArguments]
+def ofMatrix [M.IsCoxeter] : B × B → FreeGroup B :=
+  Function.uncurry fun i j =>
+    (FreeGroup.of i * FreeGroup.of j) ^ M i j
+
+/-- The set of relations corresponding to a Coxeter matrix. -/
+def toSet [M.IsCoxeter] : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix. -/
+abbrev Matrix.CoxeterGroup [M.IsCoxeter] := PresentedGroup <| CoxeterGroup.Relations.toSet M
+
+/-- A Coxeter system `CoxeterSystem W` is a structure recording the isomorphism between
+a group `W` and the group presentation corresponding to a Coxeter matrix. Equivalently, this can
+be seen as a list of generators of `W` parameterized by the underlying type of `M`, which
+satisfy the relations of the Coxeter matrix `M`. -/
+structure CoxeterSystem [M.IsCoxeter] (W : Type*) [Group W]  where
+  /-- `CoxeterSystem.ofMulEquiv` constructs a Coxeter system given an equivalence with the group
+  presentation corresponding to a Coxeter matrix `M`. -/
+  ofMulEquiv ::
+    /-- `mulEquiv` is the isomorphism between the group `W` and the group presentation
+    corresponding to a Coxeter matrix `M`. -/
+    mulEquiv : Matrix.CoxeterGroup M ≃* W
+
+/-- Coxeter system of type `M` on the group `M.CoxeterGroup`. -/
+def Matrix.coxeterSystemCoxeterGroup [M.IsCoxeter] : CoxeterSystem M M.CoxeterGroup :=
+  CoxeterSystem.ofMulEquiv (MulEquiv.refl (CoxeterGroup M))
+
+/-- A group is a Coxeter group if it is registered in a Coxeter system. -/
+class IsCoxeterGroup [M.IsCoxeter] (W : Type*) [Group W] : Prop where
+  nonempty_system : Nonempty (CoxeterSystem M W)
+
+namespace CoxeterSystem
+
+variable {W : Type*} [Group W]
+
+/-- A group registered in a Coxeter system is a Coxeter Group. -/
+lemma isCoxeterGroup [M.IsCoxeter] (cs : CoxeterSystem M W) : IsCoxeterGroup M W where
+  nonempty_system := Nonempty.intro cs
+
+instance [M.IsCoxeter] : IsCoxeterGroup M M.CoxeterGroup :=
+  isCoxeterGroup M (Matrix.coxeterSystemCoxeterGroup M)
+
+end CoxeterSystem
+
+namespace CoxeterMatrix
+
+open Matrix
+
+variable (n : ℕ)
+
+/-- The Coxeter matrix of family A(n).
+
+The corresponding Coxeter-Dynkin diagram is:
+```
+    o --- o --- o ⬝ ⬝ ⬝ ⬝ o --- o
+```
+-/
+abbrev Aₙ [NeZero n] : Matrix (Fin n) (Fin n) ℕ :=
```

Comment:
Why do you need `[NeZero n]` here and below?

Reply:
I don't need it for `Aₙ` but I do need it for `Bₙ` and `Dₙ` the way they are defined since using `Fin n` and when this is removed the error is:
```
failed to synthesize instance
  OfNat (Fin n)
```

Reply:
I don't actually, I figured out a way to get past the errors for `Bₙ` and `Dₙ` by casting 1 to `(1 : ℕ)`.